### DISCUSSION TITLE
RcppParallel compatability with checkUserInterrupt and Rcout

### DISCUSSION THREAD

**Discussion by andrewmarx** [tnagler/RcppThread#43] [Q&A]
I was wondering if RcppThread's `checkUserInterrupt()` and `Rcout` are considered safe to use in RcppParallel's threads? Apparently there is one user doing this (https://github.com/RcppCore/RcppParallel/issues/65), but I'd like to confirm they aren't just getting lucky. It's pretty essential functionality for me; even with multithreading, a lot of people using my package will want to have runs that take weeks or months, so I need to be able to give them progress updates and the ability to quit early.

On a side note, have you considered enabling Github's new Discussions feature for this repo? It would be a better place for questions like mine.

**Comment by tnagler** [comment]
Hi!

(Converted this to a discussion. Have no experience with it so far, but sounds like a good idea :wink:)

Both `RcppThread::checkUserInterrupt()` and `RcppThread::Rcout` should be safe also from  their threads. 

To prevent R from crashing, RcppThread ensures that communication with R only happens from the main thread.  In short, this is what happens:
- At the start of your program, RcppThread [stores the ID](https://github.com/tnagler/RcppThread/blob/master/inst/include/RcppThread/RMonitor.hpp#L24) of the main thread in a global variable.
-  When `checkUserInterrupt()` or `Rcout` are called, RcppThread [checks](https://github.com/tnagler/RcppThread/blob/master/inst/include/RcppThread/RMonitor.hpp#L62) whether these calls happen from the main thread. If not, there is no interruption check or printing to the console (but messages are [safely stored](https://github.com/tnagler/RcppThread/blob/master/inst/include/RcppThread/RMonitor.hpp#L93) in a buffer until they can be released).

The caveat is that the functions are useless if never called from the main thread. If I recall correctly, RcppParallel's functions always do some chunk of the work in the main thread, so this should be fine. You'll find out quickly by running a simple multithreaded example.

**Reply by andrewmarx** [reply]
Thanks, I was able to confirm that they do work with RcppParallel, and that RcppParallel does indeed make use of the main thread. The only catch is that if the main thread's partition is finished and there are no more partitions left to assign, then it can end up in a state where the main thread is idling while other threads finish up. Turns out this can be really irritating for trying to report progress to the user.
